lib/rex/proto/dns/resolver.rb

```diff
@@ -375,8 +378,99 @@ def query(name, type = Dnsruby::Types::A, cls = Dnsruby::Classes::IN)
+      entries = File.read(hf).lines.map(&:strip).select do |entry|
+        Rex::Socket.is_ip_addr?(entry.gsub(/\s+/,' ').split(' ').first) and
+        not entry.match(/::.*ip6-/) # Ignore Debian/Ubuntu-specific notation for IPv6 hosts
+      end.map do |entry|
+        entry.gsub(/\s+/,' ').split(' ')
+      end
+      entries.each do |ent|
+        next if ent.first =~ /^127\./
+        # Deal with multiple hostnames per address
+        while ent.length > 2
+          hostname = ent.pop
+          next unless MATCH_HOSTNAME.match hostname
+          begin
+            self.cache.add_static(hostname, ent.first)
+          rescue => e
+            # Deal with edge-cases in users' hostsfile
+            @logger.error(e)
+          end
+        end
+        hostname = ent.pop
+        begin
+          self.cache.add_static(hostname, ent.first) if MATCH_HOSTNAME.match hostname
+        rescue => e
+          # Deal with edge-cases in users' hostsfile
+          @logger.error(e)
+        end
+      end
+      # TODO: inotify or similar on hostsfile for live updates? Easy-button functionality
+      self.cache.start unless config[:dns_cache_no_start]
+      return
+    end
 
+    #
+    # Attempt to find answers to query in DNS cache; failing that,
+    # send remainder of DNS queries over appropriate transport and
+    # cache answers before returning to caller.
+    #
+    # @param argument
+    # @param type [Fixnum] Type of record to look up
+    # @param cls [Fixnum] Class of question to look up
+    #
+    # @return [Dnsruby::Message, nil] DNS response on success, nil on failure.
+    def send(argument, type = Dnsruby::Types::A, cls = Dnsruby::Classes::IN)
+      case argument
+      when Dnsruby::Message
+        req = argument
+      when Net::DNS::Packet, Resolv::DNS::Message
+        req = Rex::Proto::DNS::Packet.encode_drb(argument)
+      else
+        net_packet = make_query_packet(argument,type,cls)
+        # This returns a Net::DNS::Packet. Convert to Dnsruby::Message for consistency
+        req = Rex::Proto::DNS::Packet.encode_drb(net_packet)
+      end
+      resolve = req.dup
+      # Find cached items, remove request from resolved packet
+      req.question.each do |ques|
+        cached = self.cache.find(ques.qname, ques.qtype.to_s)
+        next if cached.empty?
+        req.instance_variable_set(:@answer, (req.answer + cached).uniq)
+        resolve.question.delete(ques)
+      end
+      # Resolve remaining requests, cache responses
+      if resolve.question.count > 0
+        resolved = super(resolve, type)
+        req.instance_variable_set(:@answer, (req.answer + resolved.answer).uniq)
+        resolved.answer.each do |ans|
+          self.cache.cache_record(ans)
+        end
+      end
+      # Finalize answers in response
+      # Check for empty response prior to sending
+      req.header.rcode = Dnsruby::RCode::NOERROR if req.answer.size < 1
```

Comment:
This is a bit confusing. If there is no response there is no error? Mind explaining this a bit more, its unclear why this line is needed.

Reply:
The server returns an empty response, but not because there was an error, just because there are no answers.

Reply:
Ah ok makes sense, marking resolved.